src/librustc_interface/queries.rs

```diff
@@ -283,46 +283,31 @@ impl<'tcx> Queries<'tcx> {
     }
 
     pub fn linker(&'tcx self) -> Result<Linker> {
-        let dep_graph = self.dep_graph()?;
-        let prepare_outputs = self.prepare_outputs()?;
-        let ongoing_codegen = self.ongoing_codegen()?;
+        let dep_graph = self.dep_graph()?.peek().clone();
+        let ongoing_codegen = self.ongoing_codegen()?.take();
+        let prepare_outputs = self.prepare_outputs()?.take();
 
         let sess = self.session().clone();
         let codegen_backend = self.codegen_backend().clone();
 
-        Ok(Linker {
-            sess,
-            dep_graph: dep_graph.peek().clone(),
-            prepare_outputs: prepare_outputs.take(),
-            ongoing_codegen: ongoing_codegen.take(),
-            codegen_backend,
-        })
+        let codegen_results = codegen_backend.join_codegen(ongoing_codegen, &sess, &dep_graph)?;
+        let prof = sess.prof.clone();
+        prof.generic_activity("drop_dep_graph").run(move || drop(dep_graph));
```

Comment:
This drops `dep_graph` before linking now, so this should reduce memory consumption during linking.